File: SHIPBOB_TO_GORGIAS/typescript/test/cron.test.ts
import assert from 'node:assert/strict'
import { afterEach, beforeEach, test } from 'node:test'

import * as cron from '../src/cron.js'
import { ShipBobClient } from '../src/shipbob.js'
import { GORGIAS_SECRETS, makeOrder, makePandium, recordingGorgias } from './helpers.js'

let originalLog: typeof console.log
let logged: string[]

beforeEach(() => {
    originalLog = console.log
    logged = []
    console.log = (...args: any[]) => {
        logged.push(args.join(' '))
    }
})

afterEach(() => {
    console.log = originalLog
})

/** Serves canned pages of new orders (no updated orders); records page numbers. */
class FakeShipBob implements ShipBobClient {
    pages: number[] = []
    constructor(private newPages: any[][]) {}

    async getNewOrdersPage(_cursor: Date, page: number) {
        this.pages.push(page)
        return this.newPages[page - 1] ?? []
    }
    async getUpdatedOrdersPage(_cursor: Date, _page: number) {
        return []
    }
    getUpdateDate(order: any, _cursor: Date) {
        return order.shipments[0].last_update_at
    }
}

test('clamp bounds a cursor to [now-30d, now]', () => {
    const now = new Date('2026-07-16T12:00:00Z')
    assert.equal(cron.clamp('2026-07-10T00:00:00Z', now).toISOString(), new Date('2026-07-10T00:00:00Z').toISOString())
    assert.equal(cron.clamp('2099-01-01T00:00:00Z', now).toISOString(), now.toISOString()) // future -> now
    assert.equal(
        cron.clamp(null, now).toISOString(),
        new Date(now.getTime() - cron.ONE_MONTH_MS).toISOString()
    ) // missing -> floor
})

test('run pages new orders until empty, batches one customer, advances cursor', async () => {
    const pages = [
        [
            makeOrder(1, '2026-07-05T10:00:00.1234567+00:00', { email: 'jane@example.com' }),
            makeOrder(2, '2026-07-06T10:00:00.1234567+00:00', { email: 'jane@example.com' }),
        ],
    ]
    const shipbob = new FakeShipBob(pages)
    const gorgias = recordingGorgias()

    const record = await cron.run(
        makePandium({ secrets: GORGIAS_SECRETS, config: { order_start_date: '2026-07-01' } }),
        { shipbob, gorgias }
    )

    assert.deepEqual(shipbob.pages, [1, 2]) // paged until the empty page
    assert.equal(gorgias.log.create.length, 1) // both orders batch onto one customer
    assert.equal(record.new_order_start_date, '2026-07-06T10:00:00.123456') // advanced to last order
    const finalOrders = gorgias.log.update[gorgias.log.update.length - 1][1].data.pandium.shipbob_orders
    assert.deepEqual(
        finalOrders.map((o: any) => o.id).sort(),
        [1, 2]
    )
})

test('run advances updated cursor to the oldest update seen across pages, not the last one processed', async () => {
    const ago = (days: number) => {
        const d = new Date(Date.now() - days * 86400000)
        const pad = (n: number) => String(n).padStart(2, '0')
        return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}T${pad(
            d.getUTCHours()
        )}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}.000+00:00`
    }

    class Updating implements ShipBobClient {
        pages: number[] = []
        private updatedPages = [
            [makeOrder(1, ago(2), { email: 'j@x.com' }), makeOrder(2, ago(3), { email: 'j@x.com' })],
            [makeOrder(3, ago(9), { email: 'j@x.com' }), makeOrder(4, ago(8), { email: 'j@x.com' })], // oldest update overall
            [makeOrder(5, ago(4), { email: 'j@x.com' })], // newer again, after the oldest page
        ]
        async getNewOrdersPage(_c: Date, _p: number) {
            return []
        }
        async getUpdatedOrdersPage(_cursor: Date, page: number) {
            return this.updatedPages[page - 1] ?? []
        }
        getUpdateDate(order: any, _cursor: Date) {
            return order.shipments[0].last_update_at
        }
    }

    const shipbob = new Updating()
    const gorgias = recordingGorgias()

    const record = await cron.run(
        makePandium({ secrets: GORGIAS_SECRETS, config: { order_start_date: ago(20) } }),
        { shipbob, gorgias }
    )

    assert.equal(record.updated_order_start_date, ago(9).slice(0, 23)) // not order 5, the last one processed
})

test('run flushes the partial cursor and exits zero on timeout', async () => {
    class WatchdogExit extends Error {
        constructor(public code: number) {
            super(`exit ${code}`)
        }
    }

    let capturedOnTimeout: (() => void) | undefined
    const armWatchdog = (onTimeout: () => void) => {
        capturedOnTimeout = onTimeout
        return { cancel: () => {} }
    }
    const exit = (code: number): never => {
        throw new WatchdogExit(code)
    }

    class Alarming implements ShipBobClient {
        async getNewOrdersPage(_cursor: Date, page: number) {
            if (page === 1) {
                return [makeOrder(1, '2026-07-05T10:00:00.1234567+00:00', { email: 'j@x.com' })]
            }
            capturedOnTimeout?.() // trip the watchdog before page 2 is processed
            return [makeOrder(2, '2026-07-06T10:00:00.1234567+00:00', { email: 'j@x.com' })]
        }
        async getUpdatedOrdersPage(_cursor: Date, _page: number) {
            return []
        }
        getUpdateDate(order: any, _cursor: Date) {
            return order.shipments[0].last_update_at
        }
    }

    const shipbob = new Alarming()
    const gorgias = recordingGorgias()

    await assert.rejects(
        cron.run(makePandium({ secrets: GORGIAS_SECRETS, config: { order_start_date: '2026-07-01' } }), {
            shipbob,
            gorgias,
            armWatchdog,
            exit,
        }),
        (err: any) => {
            assert.ok(err instanceof WatchdogExit)
            assert.equal(err.code, 0) // timed-out run still succeeds so progress is merged
            return true
        }
    )

    const flushed = JSON.parse(logged[logged.length - 1])
    assert.equal(flushed.new_order_start_date, '2026-07-05T10:00:00.123456') // only order 1 processed
})

File: shipbob_to_gorgias/ruby/test/cron_test.rb
require_relative 'test_helper'
require_relative 'fakes'
require_relative '../cron'
require 'stringio'
require 'time'

class CronTest < Minitest::Test
  # Ruby's `loop do ... end` silently catches StopIteration and exits the loop normally, so a
  # sentinel used to escape Cron.run's paging loops from a test must be some other class.
  class SimulatedExit < StandardError; end

  def ago(days)
    (Time.now.utc - (days * 24 * 60 * 60)).strftime('%Y-%m-%dT%H:%M:%S') + '.1234567+00:00'
  end

  def test_clamp_bounds_cursor_between_one_month_ago_and_now
    now = Time.utc(2026, 7, 16, 12, 0, 0)
    assert_equal Time.utc(2026, 7, 10), Sb2Gorgias::Cron.clamp('2026-07-10T00:00:00', now)
    assert_equal now, Sb2Gorgias::Cron.clamp('2099-01-01T00:00:00', now)
    assert_equal now - Sb2Gorgias::Cron::ONE_MONTH_SECONDS, Sb2Gorgias::Cron.clamp(nil, now)
  end

  def test_upsert_replaces_in_place_without_resort
    orders = [{ 'id' => 5 }, { 'id' => 1 }] # deliberately unsorted
    result = Sb2Gorgias::Cron.upsert(orders, { 'id' => 1, 'updated' => true }, false)
    assert_equal [{ 'id' => 5 }, { 'id' => 1, 'updated' => true }], result # order preserved, not re-sorted
  end

  def test_upsert_appends_sorts_by_id_and_trims
    orders = (1..10).map { |i| { 'id' => i } }
    result = Sb2Gorgias::Cron.upsert(orders, { 'id' => 11 }, true)
    assert_equal 10, result.size
    assert_equal 11, result.first['id']
    assert_equal 2, result.last['id'] # id 1 trimmed off
  end

  def test_run_pages_until_empty_upserts_customer_and_advances_cursor
    shipbob = FakeShipBobClient.new(
      new_pages: [[make_order(1, ago(6), email: 'jane@example.com'), make_order(2, ago(5), email: 'jane@example.com')]],
      updated_pages: []
    )
    gorgias = recording_gorgias
    pandium = make_pandium(secrets: GORGIAS_SECRETS, config: { 'order_start_date' => ago(20) })

    record = Sb2Gorgias::Cron.run(pandium, shipbob: shipbob, gorgias: gorgias, now: Time.now.utc)

    assert_equal [1, 2], shipbob.pages['new'] # paged until the empty page confirmed exhaustion
    assert_equal 1, gorgias.log[:create].size # both orders batch onto one customer
    assert_equal ago(5)[0, 26], record['new_order_start_date']

    final_orders = gorgias.log[:update].last[1]['data']['pandium']['shipbob_orders']
    assert_equal [1, 2], final_orders.map { |o| o['id'] }.sort
  end

  def test_run_advances_updated_cursor_to_oldest_update_across_pages
    shipbob = FakeShipBobClient.new(
      new_pages: [],
      updated_pages: [
        [make_order(1, ago(2), email: 'j@x.com'), make_order(2, ago(3), email: 'j@x.com')],
        [make_order(3, ago(9), email: 'j@x.com'), make_order(4, ago(8), email: 'j@x.com')],
        [make_order(5, ago(4), email: 'j@x.com')],
      ]
    )
    pandium = make_pandium(secrets: GORGIAS_SECRETS, config: { 'order_start_date' => ago(20) })

    record = Sb2Gorgias::Cron.run(pandium, shipbob: shipbob, gorgias: recording_gorgias, now: Time.now.utc)

    # get_update_date passes the raw shipment timestamp through untouched (only truncated to
    # 23 chars), so the expected value is a plain slice of the fixture string, not a
    # parsed-and-reformatted one.
    assert_equal ago(9)[0, 23], record['updated_order_start_date'] # not order 5, the last processed
  end

  # A watchdog firing mid-run is still a success: it flushes the completed half's progress and
  # leaves the interrupted half at its starting cursor, the same success path as a run that
  # finishes on its own.
  def test_watchdog_flushes_the_finished_half_and_leaves_the_interrupted_one
    start = ago(20)
    captured_timeout = nil
    arm_watchdog = lambda do |_seconds, &on_timeout|
      captured_timeout = on_timeout
      -> {} # cancel: no-op, nothing to test here
    end

    shipbob = FakeShipBobClient.new(
      new_pages: [[make_order(1, ago(6), email: 'j@x.com')]],
      updated_pages: [
        [make_order(2, ago(2), email: 'j@x.com')],
        [make_order(3, ago(9), email: 'j@x.com')], # never read
      ],
      on_page: lambda do |half, page|
        captured_timeout.call if half == 'updated' && page == 2
      end
    )
    pandium = make_pandium(secrets: GORGIAS_SECRETS, config: { 'order_start_date' => start })

    original_stdout = $stdout
    $stdout = StringIO.new
    exit_code = nil
    begin
      Sb2Gorgias::Cron.run(pandium, shipbob: shipbob, gorgias: recording_gorgias, arm_watchdog: arm_watchdog,
                                     exit_fn: ->(code) { exit_code = code; raise SimulatedExit },
                                     now: Time.now.utc)
    rescue SimulatedExit
      nil # the watchdog fired mid-page and stopped the run via the injected exit_fn
    ensure
      flushed = JSON.parse($stdout.string.strip)
      $stdout = original_stdout
    end

    assert_equal 0, exit_code # a timed-out run still counts as successful
    assert_equal ago(6)[0, 26], flushed['new_order_start_date'] # that half finished
    assert_equal start[0, 26], flushed['updated_order_start_date'] # this one did not
  end
end
